ComponentKit/Core/Scope/CKDetectComponentScopeCollisions.mm

```diff
@@ -31,9 +32,69 @@
 }
 #pragma clang diagnostic pop
 
-void CKDetectComponentScopeCollisions(const CKComponentLayout &layout)
+static CKComponent * findCollisionComponent(CKComponent *component,
+                                            id<NSObject> collisionScope,
+                                            NSMapTable<CKComponent *, CKComponent*> *componentsToParentComponents)
+{
+  for(CKComponent* componentKey in componentsToParentComponents){
+    const id<NSObject> scopeFrameToken = [componentKey scopeFrameToken];
+    // The colission node has the same scopeToken but it's not this node.
+    if([collisionScope isEqual:scopeFrameToken] && ![component isEqual:componentKey]){
+      return componentKey;
+    }
+  }
+  return nil;
+}
+
+NSUInteger heightOfComponent(CKComponent *component,
+                             NSMapTable<CKComponent *, CKComponent*> *componentsToParentComponents)
+{
+  NSUInteger heightOnTree = 0;
+  CKComponent *parentComponent = [componentsToParentComponents objectForKey:component];
+  while (parentComponent) {
+    parentComponent = [componentsToParentComponents objectForKey:parentComponent];
+    heightOnTree += 1;
+  }
+  return heightOnTree;
+}
+
+static CKComponent* lowestCommonAncestor(CKComponent *component,
+                                         id<NSObject> collisionScope,
+                                         NSMapTable<CKComponent *, CKComponent*> *componentsToParentComponents)
+{
+  // First we need to find the node we're having a collision with.
+  CKComponent *component2 = findCollisionComponent(component, collisionScope, componentsToParentComponents);
+  
+  // Then we find the How deep each node is
+  NSUInteger heightNode1 = heightOfComponent(component, componentsToParentComponents);
+  NSUInteger heightNode2 = heightOfComponent(component2, componentsToParentComponents);
+  
+  if (heightNode1 > heightNode2) {
+    std::swap(heightNode1, heightNode2);
+    std::swap(component, component2);
+  }
+  
+  // Just making sure both paths start from the same deep level
+  while (heightNode1 < heightNode2) {
+    if (!component2)
+      return nil;
+    component2 = [componentsToParentComponents objectForKey:component2];
+    heightNode2 -= 1;
+  }
+  
+  // Just comparing the paths until we find the lowest common ancestor
```

Comment:
This is debugging code so we can get away with being a little less efficient in favor of clarity.